dll/win32/imm32/imm.c

```diff
@@ -1439,22 +1439,25 @@ DWORD WINAPI ImmGetCandidateListW(
 BOOL WINAPI ImmGetCandidateWindow(
   HIMC hIMC, DWORD dwIndex, LPCANDIDATEFORM lpCandidate)
 {
-    InputContextData *data = get_imc_data(hIMC);
-
-    TRACE("%p, %d, %p\n", hIMC, dwIndex, lpCandidate);
-
-    if (!data || !lpCandidate)
-        return FALSE;
+    BOOL ret = FALSE;
+    LPINPUTCONTEXT pIC;
+    LPCANDIDATEFORM pCF;
 
-    if (dwIndex >= ARRAY_SIZE(data->IMC.cfCandForm))
-        return FALSE;
+    TRACE("ImmGetCandidateWindow(%p, %lu, %p)\n", hIMC, dwIndex, lpCandidate);
 
-    if (data->IMC.cfCandForm[dwIndex].dwIndex != dwIndex)
+    pIC = ImmLockIMC(hIMC);
+    if (pIC  == NULL)
         return FALSE;
 
-    *lpCandidate = data->IMC.cfCandForm[dwIndex];
+    pCF = &pIC->cfCandForm[dwIndex];
+    if (pCF->dwIndex != 0xFFFFFFFF)
```

Comment:
`0xFFFFFFFF == -1` but what is that supposed to represent actually? Could you please add some comment explaining what is this or at least define this value?

Reply:
Why are there two different indices in play? Also what could be the expected value of `pCF->dwIndex` in case it's not == -1 ?

Reply:
Japanese input uses the "candidate list" to choose the string from the list.

`INPUTCONTEXT` Structure:
https://katahiromz.web.fc2.com/colony3rd/imehackerz/en/INPUTCONTEXT.html
`CANDIDATEFORM` Structure:
https://katahiromz.web.fc2.com/colony3rd/imehackerz/en/CANDIDATEFORM.html

`ImmGetCandidateWindow`'s 2nd parameter `dwIndex` is the index of a candidate window information (`CANDIDATEFORM`) of `INPUTCONTEXT.cfCandForm[4]`. There can be multiple candidate windows (up to 4).

`CANDIDATEFORM.dwIndex` is a candidate list identifier. A candidate list identifier is the index of the `CANDIDATELIST` of the memory block `CANDIDATEINFO`. The `CANDIDATEINFO` structure can be accessed from `INPUTCONTEXT.hCandInfo` via `ImmLockIMCC/ImmUnlockIMCC`.

`CANDIDATEINFO` is a continuous memory block and contains the offsets (`DWORD dwOffset[32];`) to each list (`CANDIDATELIST`).

The candidate list can be empty. If so, `pCF->dwIndex` is `0xFFFFFFFF`.

Reply:
Then use `ULONG_MAX`, or `(ULONG)(-1)`

Reply:
OK.

Reply:
This is the candidate window:
![image](https://user-images.githubusercontent.com/2107452/125595130-0c44b755-e404-4078-84c1-3cab6be91e1a.png)